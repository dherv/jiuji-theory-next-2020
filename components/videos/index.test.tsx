import React from "react";
import { render, cleanup, wait, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Videos from ".";
import { MockedComponentClass } from "react-dom/test-utils";

describe("Videos component", () => {
  afterEach(() => {
    cleanup();
    jest.resetAllMocks();
  });
  const props = {};
  const fetch = () =>
    jest.fn().mockImplementation((url: String) => {
      const fetch = Promise.resolve({
        ok: true,
        json() {
          return [
            { youtubeId: "1", title: "test1", description: "testDescription1" },
            { youtubeId: "2", title: "test2", description: "testDescription2" }
          ];
        }
      });
      return fetch;
    });

  window.fetch = fetch();
  test("should fetch and display a list of videos ", async () => {
    const { asFragment, getByText } = render(<Videos />);
    await wait(() => {
      expect(getByText("test1")).toBeDefined();
      expect(getByText("test2")).toBeDefined();
      expect(window.fetch).toHaveBeenCalledTimes(1);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  test("should fetch videos on search and replace the default list of videos", async () => {
    window.fetch = fetch();
    const { asFragment, getByRole } = render(<Videos />);

    await wait();
    expect(window.fetch).toHaveBeenCalledTimes(1);
    const firstRender = asFragment();

    fireEvent.click(getByRole("button"));
    await wait(() => {
      expect((window.fetch as jest.Mock).mock.calls[0][0]).toBe(
        "http://localhost:9000/videos"
      );
      expect((window.fetch as jest.Mock).mock.calls[1][0]).toBe(
        "http://localhost:9000/videos/search"
      );
    });

    expect(firstRender).not.toEqual(asFragment());
    expect(asFragment()).toMatchSnapshot();
  });
});
