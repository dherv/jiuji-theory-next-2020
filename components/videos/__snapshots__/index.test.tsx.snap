// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Videos component should fetch and display a list of videos  1`] = `
<DocumentFragment>
  <form>
    <label
      for="searchInput"
    >
      search BJJ videos
    </label>
    <input
      id="searchInput"
      placeholder="search"
    />
    <button>
      search
    </button>
  </form>
  <ul
    class="VideosListstyled__List-sc-599crm-0 iuxoSV"
  >
    <li
      class="VideosListstyled__ListItem-sc-599crm-1 fsPLJX"
    >
      <div
        class="VideosListstyled__VideoCard-sc-599crm-2 gjvOAZ"
      >
        <div
          class=""
        >
          <div />
        </div>
        <div
          class="VideosListstyled__VideoCardContent-sc-599crm-3 eFwPQv"
        >
          <h5
            class="VideosListstyled__VideoCardTitle-sc-599crm-4 beRIiU"
          >
            test1
          </h5>
          <h5>
            testDescription1
          </h5>
        </div>
      </div>
    </li>
    <li
      class="VideosListstyled__ListItem-sc-599crm-1 fsPLJX"
    >
      <div
        class="VideosListstyled__VideoCard-sc-599crm-2 gjvOAZ"
      >
        <div
          class=""
        >
          <div />
        </div>
        <div
          class="VideosListstyled__VideoCardContent-sc-599crm-3 eFwPQv"
        >
          <h5
            class="VideosListstyled__VideoCardTitle-sc-599crm-4 beRIiU"
          >
            test2
          </h5>
          <h5>
            testDescription2
          </h5>
        </div>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Videos component should fetch videos on search and replace the default list of videos 1`] = `
<DocumentFragment>
  <form>
    <label
      for="searchInput"
    >
      search BJJ videos
    </label>
    <input
      id="searchInput"
      placeholder="search"
    />
    <button>
      search
    </button>
  </form>
  <ul
    class="VideosListSearchstyled__List-f1y4wp-0 kQuvvS"
  >
    <li
      class="VideosListSearchstyled__ListItem-f1y4wp-1 hFYOFR"
    >
      <div
        class="VideosListSearchstyled__VideoCard-f1y4wp-2 ioVfgg"
      >
        <div
          class=""
        >
          <div />
        </div>
        <div
          class="VideosListSearchstyled__VideoCardContent-f1y4wp-3 fXMJsV"
        >
          <h5
            class="VideosListSearchstyled__VideoCardTitle-f1y4wp-4 lkkzpL"
          >
            test1
          </h5>
          <h5>
            testDescription1
          </h5>
        </div>
      </div>
    </li>
    <li
      class="VideosListSearchstyled__ListItem-f1y4wp-1 hFYOFR"
    >
      <div
        class="VideosListSearchstyled__VideoCard-f1y4wp-2 ioVfgg"
      >
        <div
          class=""
        >
          <div />
        </div>
        <div
          class="VideosListSearchstyled__VideoCardContent-f1y4wp-3 fXMJsV"
        >
          <h5
            class="VideosListSearchstyled__VideoCardTitle-f1y4wp-4 lkkzpL"
          >
            test2
          </h5>
          <h5>
            testDescription2
          </h5>
        </div>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
